pkgs/pkgs-lib/formats/hocon/src/src/main.rs

```diff
@@ -0,0 +1,234 @@
+use serde_json::{Map, Value};
+
+#[derive(Debug)]
+enum HOCONValue {
+    Null,
+    Append(Box<HOCONValue>),
+    Bool(bool),
+    Int(i64),
+    Float(f64),
+    String(String),
+    List(Vec<HOCONValue>),
+    Substitution(String, bool),
+    Object(Vec<HOCONInclude>, Vec<(String, HOCONValue)>),
+}
+
+#[derive(Debug)]
+enum HOCONInclude {
+    Heuristic(String, bool),
+    Url(String, bool),
+    File(String, bool),
+    ClassPath(String, bool),
+}
+
+impl HOCONInclude {
+    fn map_fst(&self, f: &dyn Fn(&String) -> String) -> HOCONInclude {
+        match self {
+            HOCONInclude::Heuristic(s, r) => HOCONInclude::Heuristic(f(s), *r),
+            HOCONInclude::Url(s, r) => HOCONInclude::Url(f(s), *r),
+            HOCONInclude::File(s, r) => HOCONInclude::File(f(s), *r),
+            HOCONInclude::ClassPath(s, r) => HOCONInclude::ClassPath(f(s), *r),
+        }
+    }
+}
+
+fn escape_string(s: &str) -> String {
+    s.replace('\\', "\\\\").replace('\"', "\\\"").replace('\n', "\\n").replace('\t', "\\t")
+}
```

Comment:
Would be great to have a link listing that only these characters need to be escaped

Reply:
I don't think there's a single place in the docs that points this out. I believe these were my reasons for the escapes. Should I add it as a comment?

- `\` is there for ensuring existing escapes are correct.
- `"` is there because the entire string will get wrapped.
- `\n` is there because wrapping the string with `"` implies a single line string.
- `\t` might not even be necessary. Maybe I should remove it?

Reply:
Oh actually it does mention this:

> - quoted strings are in the same format as JSON strings

So here it also seems like we should be able to just use `serde_json`'s string encoding

Reply:
I've fixed it up. I discovered that it's a bit ambiguous how substitutions should be escaped, if at all, so I've left it unescaped for now. Hoping that the `checkPhase` catches any invalid configuration left.